### PR TITLE
[Graphbolt] Graphbolt Link Prediction Example

### PR DESCRIPTION
**Pull request by keli-wen**
## Description

A draft for Graphbolt link-perdiction example.

## How to make the ogbl_citation2 dataset?

Utilize the example provided in https://ogb.stanford.edu/docs/linkprop/ to download the ogbl-citation2 dataset.

```python
import torch
from ogb.linkproppred import LinkPropPredDataset

dataset_name = "ogbl-citation2"
# Set the download directory
data_root = "./dataset" 

# Download.
dataset = LinkPropPredDataset(name=dataset_name, root=data_root)
```

After running the code above, navigate to the respective dataset folder and look for `ogbl_citation2`; all the data you need can be found there. Below is the `metadata.yaml` file we're currently using:

```yaml
dataset_name: ogbl_citation2 
graph:
  nodes:
    - num: 2927963
  edges:
    - format: csv
      path: edges/cite.csv
  feature_data:
feature_data:
  - domain: node
    type: null
    name: feat
    format: numpy
    in_memory: true
    path: data/node-feat.npy
  - domain: node
    type: null
    name: year
    format: numpy
    in_memory: true
    path: data/node-year.npy
tasks:
  - name: "link_prediction"
    num_classes: 2
    train_set:
      - type_name: null
        data:
        # (n, 2)
        - name: node_pairs
          format: numpy
          path: set/train_node_pairs.npy
          in_memory: true
    validation_set:
      - type_name: null
        data:
        - name: node_pairs
          format: numpy
          path: set/valid_node_pairs.npy
          in_memory: true
        - name: negative_dsts
          format: numpy
          path: set/valid_negative_dsts.npy
          in_memory: true
    test_set:
      - type_name: null
        data:
        - name: node_pairs
          format: numpy
          path: set/test_node_pairs.npy
          in_memory: true
        - name: negative_dsts
          format: numpy
          path: set/test_negative_dsts.npy
          in_memory: true
```

You'll need to convert the **raw dataset** into the corresponding structure and organize it into any folder of your choice. The final file structure should look like this:

```
.
├── data
│   ├── node-feat.npy
│   └── node-year.npy
├── edges
│   └── cite.csv
├── metadata.yaml
└── set
    ├── test_negative_dsts.npy
    ├── test_node_pairs.npy
    ├── train_node_pairs.npy
    ├── valid_negative_dsts.npy
    └── valid_node_pairs.npy
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: link_prediction.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+def main(args):
+    if not torch.cuda.is_available():
+        args.mode = "cpu"
+    print(f"Training in {args.mode} mode.")
+
+    # Load and preprocess dataset.
+    print("Loading data")
+    dataset = gb.OnDiskDataset(
+        "/home/ubuntu/wklwork/work#4/example_ogbl_citation2"
```
Comment (anko-intel) [inline link_prediction.py:222]: Hi, I would like to run the example myself. Could you please attach the YAML file and a description of how to prepare the data set for it?

Comment (keli-wen) [reply link_prediction.py:222]: Thank you for your attention✨. This example is currently a provisional version due to potential changes in many of Graphbolt's APIs. For now, it serves as an internal reference. We plan to merge it into the master branch in the future and will provide an official channel for dataset downloading and processing. @frozenbugs

Comment (frozenbugs) [reply link_prediction.py]: Yes, @keli-wen once it is ready, please also draft readme so that ppl can run it by themselves. Thanks @anko-intel for bringing it up.

**Comment by mfbalin** [comment]
After #6214 and #6203 are merged, can we add a `--labor` option to this script so that we can test gb.LayerNeighborSampler as well?

**Comment by frozenbugs** [comment]
> After #6214 and #6203 are merged, can we add a `--labor` option to this script so that we can test `gb.LayerNeighborSampler` as well?

How about we create an additional example specifically focusing on the labor sampler? It would be beneficial to have an example demonstrating Labor Sampling with Hetero Node Classification. We could place this example in the "examples/advanced/labor_sampling/" directory. This way, if there are more related examples like GPUCache in the future, they can all be organized within the same folder.

**Comment by mfbalin** [comment]
I feel like that would be hiding LayerNeighbor from users somehow. Anywhere NeighborSampler is used, it is possible to have a `--labor` flag. This way, users can compare the different samplers in each scenario NeighborSampler is used. It would also highlight that Graphbolt supports LayerNeighbor as a first class citizen. This way if any issues come up, we can address them. We could also create advanced examples to demonstrate the advanced Labor features, here no advanced features are used, simply drop-in replacement.

@frozenbugs What do you think? LayerNeighbor is still Neighbor afterall. I would more than appreciate the gesture if we gave the `--labor` option more visibility. Users don't know about this alternative at all and what they are missing by not using it.

File: examples/sampling/graphbolt/link_prediction.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+def train(args, graph, features, train_set, valid_set, model):
+    total_loss = 0.0
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
+    dataloader = get_dataloader(args, graph, features, train_set)
+
+    for epoch in tqdm.trange(args.epochs):
+        model.train()
+        for step, data in enumerate(dataloader):
+            # Unpack LinkPredictionBlock.
+            labels = data.label.float()
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/link_prediction.py:151]: where is `data.label` is assigned? data_block_converter?

Comment (keli-wen) [reply examples/sampling/graphbolt/link_prediction.py:151]: In `NegativeSampler._collate` https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/negative_sampler.py#L117-L121.

File: examples/sampling/graphbolt/link_prediction.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+    # [Role]:
+    # Initialize the UniformNegativeSampler for negative sampling in link
+    # prediction.
+    ############################################################################
+    negative_sampler = gb.UniformNegativeSampler(
+        data_block_converter,
+        args.neg_ratio,
+        format,
+        graph,
+    )
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/link_prediction.py:171]: When the is_train is false, it seems that the set (to_link_block_eval) already contains negative edges, why do we still need negative sampler?

**Comment by frozenbugs** [comment]
Overall the example is in a good shape now. Let's wait for the final clean up of the graphbolt UX, and have this example merged.

File: examples/sampling/graphbolt/link_prediction.py (hunk 3 above)
Comment (keli-wen) [reply examples/sampling/graphbolt/link_prediction.py:171]: if the is_train not true, subgraph sampler will directly use `data_block_convert`. Maybe we should not initialize the `negative _sampler` if in eval mode?

**Comment by keli-wen** [comment]
> Overall the example is in a good shape now. Let's wait for the final clean up of the graphbolt UX, and have this example merged.

Perhaps it would be better for me to put the contents of the README in the **description** of this PR? Because we will provide official channels in the future? Then this README is no longer needed.

File: examples/sampling/graphbolt/README.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
+## How to make your dataset?
+
+Utilize the example provided in https://ogb.stanford.edu/docs/linkprop/ to download the ogbl-citation2 dataset.
+
+```python
+import torch
+from ogb.linkproppred import LinkPropPredDataset
+
+dataset_names = "ogbl-citation2"
```
Comment (anko-intel) [inline examples/sampling/graphbolt/README.md:9]:
```suggestion
dataset_name = "ogbl-citation2"
```

**Comment by frozenbugs** [comment]
> > Overall the example is in a good shape now. Let's wait for the final clean up of the graphbolt UX, and have this example merged.
> 
> Perhaps it would be better for me to put the contents of the README in the **description** of this PR? Because we will provide official channels in the future? Then this README is no longer needed.

We still need README, and it is better to describe the data structure of the example. Let's discuss offline to clarify what needs to be included in README.

File: examples/sampling/graphbolt/link_prediction.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+training pipeline using GraphBolt.
+
+While node classification predicts labels for nodes based on their
+local neighborhoods, link prediction assesses the likelihood of an edge
+existing between two nodes, necessitating different sampling strategies
+that account for pairs of nodes and their joint neighborhoods.
+
+Before reading this example, please familiar yourself with graphsage node
+classification by reading the example in the
+`examples/core/graphsage/link_prediction.py`
```
Comment (anko-intel) [inline examples/sampling/graphbolt/link_prediction.py:19]: there is no such file

File: examples/sampling/graphbolt/link_prediction.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+def main(args):
+    if not torch.cuda.is_available():
+        args.mode = "cpu"
+    print(f"Training in {args.mode} mode.")
+
+    # Load and preprocess dataset.
+    print("Loading data")
+    dataset = gb.OnDiskDataset(
+        "/home/ubuntu/wklwork/work#4/example_ogbl_citation2"
```
Comment (Rhett-Ying) [inline examples/sampling/graphbolt/link_prediction.py:357]: should be replaced by argument?

File: examples/sampling/graphbolt/link_prediction.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+        hidden_x = x
+        for layer_idx, (layer, block) in enumerate(zip(self.layers, blocks)):
+            hidden_x = layer(block, hidden_x)
+            is_last_layer = layer_idx == len(self.layers) - 1
+            if not is_last_layer:
+                hidden_x = F.relu(hidden_x)
+        return hidden_x
+
+
+def current_dataloader(args, graph, features, itemset, is_train=True):
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/link_prediction.py:82]:
```suggestion
def create_dataloader(args, graph, features, itemset, is_train=True):
```

**Review by frozenbugs — APPROVED** [review]

File: examples/sampling/graphbolt/link_prediction.py (hunk 7 above)
Comment (keli-wen) [reply examples/sampling/graphbolt/link_prediction.py]: Done
